crates/oxc_transformer/src/es2017/async_to_generator.rs

```diff
@@ -0,0 +1,171 @@
+//!   await bar();
+//! }
+//! const foo2 = async () => {
+//!   await bar();
+//! };
+//! ```
+//!
+//! Output:
+//! ```js
+//! function foo() {
+//!   return _asyncToGenerator(this, null, function* () {
+//!     yield bar();
+//!   })
+//! }
+//! const foo2 = () => _asyncToGenerator(this, null, function* () {
+//!   yield bar();
+//! }
+//! ```
+//!
+//! ## Implementation
+//!
+//! Implementation based on [@babel/plugin-transform-async-to-generator](https://babel.dev/docs/babel-plugin-transform-async-to-generator) and [esbuild](https://github.com/evanw/esbuild/blob/main/internal/js_parser/js_parser_lower.go#L392).
+//!
+//!
+//! Reference:
+//! * Babel docs: <https://babeljs.io/docs/en/babel-plugin-transform-async-to-generator>
+//! * Esbuild implementation: <https://github.com/evanw/esbuild/blob/main/internal/js_parser/js_parser_lower.go#L392>
+//! * Babel implementation: <https://github.com/babel/babel/blob/main/packages/babel-plugin-transform-async-to-generator>
+//! * Babel helper implementation: <https://github.com/babel/babel/blob/main/packages/babel-helper-remap-async-to-generator>
+//! * Async / Await TC39 proposal: <https://github.com/tc39/proposal-async-await>
+//!
+
+use crate::context::Ctx;
+use oxc_allocator::CloneIn;
+use oxc_ast::ast::{ArrowFunctionExpression, Expression, FormalParameterKind, Function, FunctionType, Statement, YieldExpression};
+use oxc_ast::NONE;
+use oxc_span::{Atom, SPAN};
+use oxc_syntax::reference::ReferenceFlags;
+use oxc_syntax::symbol::SymbolId;
+use oxc_traverse::{Ancestor, Traverse, TraverseCtx};
+
+pub struct AsyncToGenerator<'a> {
+    _ctx: Ctx<'a>,
+}
+
+impl<'a> AsyncToGenerator<'a> {
+    pub fn new(ctx: Ctx<'a>) -> Self {
+        Self { _ctx: ctx }
+    }
+
+    fn get_helper_callee(symbol_id: Option<SymbolId>, ctx: &mut TraverseCtx<'a>) -> Expression<'a> {
+        let ident = ctx.create_reference_id(
+            SPAN,
+            Atom::from("babelHelpers"),
+            symbol_id,
+            ReferenceFlags::Read,
+        );
+        let object = ctx.ast.expression_from_identifier_reference(ident);
+        let property = ctx.ast.identifier_name(SPAN, Atom::from("asyncToGenerator"));
+        Expression::from(ctx.ast.member_expression_static(SPAN, object, property, false))
+    }
+}
+
+impl<'a> Traverse<'a> for AsyncToGenerator<'a> {
+    fn exit_expression(&mut self, expr: &mut Expression<'a>, ctx: &mut TraverseCtx<'a>) {
+        if let Expression::AwaitExpression(await_expr) = expr {
+            // Do not transform top-level await
+            if ctx.ancestry.ancestors().any(|ancestor| {
+                matches!(
+                    ancestor,
+                    Ancestor::FunctionBody(_) | Ancestor::ArrowFunctionExpressionBody(_)
+                )
+            }) {
+                let yield_expression = YieldExpression {
+                    span: SPAN,
+                    delegate: false,
+                    argument: Some(await_expr.argument.clone_in(ctx.ast.allocator)),
```

Comment:
I'm guessing this line is where some of the poor performance is coming from. There's no need to clone.

```rs
argument: Some(ctx.ast.move_expression(&mut await_expr.argument)),
```

`move_expression` just takes the `Expression`, without cloning it.